graph/encoding/graphql: add new package for decoding GraphQL output

kortschak opened the pull request:
Please take a look.

Depends on #168.
Closes #72.

mewmew commented on graph/encoding/graphql/decode.go:
```diff
@@ -0,0 +1,157 @@
+	// uidName is the name of the UID field in the source JSON.
+	uidName string
+
+	nodes map[string]graph.Node
```
Add a comment stating the key of the map; e.g.

```go
// nodes maps from node value to graph node.
```

Is this comment correct? Are node values unique?

kortschak replied:
Done.

UIDs are unique in the input text.

mewmew commented on graph/encoding/graphql/decode.go:
```diff
@@ -0,0 +1,157 @@
+			if err != nil {
+				return err
+			}
+			value := fmt.Sprint(v)
```
What does the call to `fmt.Sprint` do? Can it be replaced with a type assertion for `fmt.Stringer`?

kortschak replied:
Everything can be represented as a string, `fmt.Sprint(v)` does this conversion. The two things we are likely to get here are a `string` or an `int`. Neither of these implement `fmt.Stringer`.

mewmew replied:
Ok.

mewmew commented on graph/encoding/graphql/decode.go:
```diff
@@ -0,0 +1,157 @@
+			return err
+		}
+		if attr == g.uidName {
+			value := fmt.Sprint(v)
```
Same as comment above.

kortschak replied:
Same answer.

kortschak commented:
PTAL

mewmew commented:
LGTM

kortschak commented:
Can you re-approve please.

mewmew commented:
I tried to have a look around, but can't find the option to do so in the web UI.

![Where to re-approve?](http://i.imgur.com/4xbDTTA.png)

mewmew commented:
It says

>  kortschak dismissed mewmew’s stale review via 1b863d0 7 hours ago

and now, the review button is gone for what it looks like.

mewmew commented:
I think the reason may be the force push,

> We went looking everywhere, but couldn’t find those commits.

kortschak commented:
You need to be in the [files changed tab](https://github.com/gonum/gonum/pull/169/files), where you'll see green "Review changes" button/menu/dialogue. The review is dismissed by any additional push.

mewmew commented:
> You need to be in the files changed tab, where you'll see green "Review changes" button/menu/dialogue. The review is dismissed by any additional push.

Ok, I'll take a look.

mewmew commented:
For future notes,

@kortschak Did you do a force push, or how did you push the latest incremental commit?

The benefit of uploading incremental commits, is that it is easier to review the changes, then you see the exact lines which change. Also, when the PR is complete, you may hit [squash and merge](https://github.com/blog/2141-squash-your-commits) to turn those several commits into a single one.

kortschak commented:
It was forced. In this PR there was very minimal change (the changes are visible from the GUI - it's unfortunate githubs git GC is so aggressive in removing those commits - it did not use to be) in addressing the comments - and significantly, the PR would not pass unless rebased onto the changes in #168 merged into master. Normally I do do incremental and retain the history (at least since the addition of the squash option on the GUI).

mewmew commented:
I see. Thanks for the background.
